Update Blog Post “how-to-send-sms-messages-with-node-js-and-express-dr” (#1601)
Nexmo/deved-platform: content/blog/en/how-to-send-sms-messages-with-node-js-and-express-dr.md

---
title: Send SMS Messages with Node.js and Express using the Vonage SMS API
description: A step-by-step tutorial on how to send SMS messages with Node.js

>   and Express using the Nexmo SMS API and Node.js client library.

Reviewer: ```suggestion
  and Express using the Vonage SMS API and Node.js client library.
```

**View** **[the source code on GitHub](https://github.com/nexmo-community/send-sms-nodejs-express)**

### Prerequisites

Before starting this tutorial, make sure you have:

- the basic understanding of JavaScript and Node.js
- [Node.js](https://nodejs.org/en/) installed on your machine

<sign-up number></sign-up>

## Using the Vonage Server SDK for Node.js

First, run `npm init` in your project folder, then use npm to install `@vonage/server-sdk`, the Server SDK for Node.js in your working directory:

`$ npm install @vonage/server-sdk --save`

Add `"type": "module"` to your `package.json` file to be able to use [import](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/import) statements.

Create a `.js` file, let’s call it `index.js`, and in the file, initialize a `Vonage` instance with your credentials. Find your API key and secret in your [Vonage Dashboard](https://dashboard.nexmo.com/).

```javascript
import Vonage from "@vonage/server-sdk";

const vonage = new Vonage({
	apiKey: VONAGE_API_KEY,
	apiSecret: VONAGE_API_SECRET,
});
```

> There are also optional params you can use. Find out more in [the SMS API Reference](https://developer.vonage.com/api/sms?theme=dark).

Reviewer: I think something may have happened with copy and paste.
```suggestion
There are also optional params such as, `applicationId`, `privateKey`, and `options` object. You can find out more on [the **vonage-node-sdk** library repo on GitHub](https://github.com/vonage/vonage-node-sdk).
```

Reviewer 2: Not really, just the previous information was incorrect. I found the API reference to be a better source to find out more about the available parameters.

> ## Send SMS Messages with Node.js and the Vonage SMS API

Reviewer: Small thing, but the code is sending one message and not multiple ones. I'm cool with either way.
```suggestion
## Send an SMS Message with Node.js and the Vonage SMS API
```

To send a message, use the `vonage.message.sendSms` function and pass your virtual number you are sending the message from, a recipient number, and the message to be sent.

Also, you can pass [optional params](https://developer.vonage.com/api/sms?theme=dark), and a callback.

Let's hard-code the phone number (which should start with a country code, e.g. "15105551234") and a message for now to try the API. Replace `YOUR_VIRTUAL_NUMBER` with one of your [Vonage numbers](https://dashboard.nexmo.com/your-numbers).

```javascript
vonage.message.sendSms(
	"YOUR_VIRTUAL_NUMBER",
	"15105551234",
	"yo",
	(err, responseData) => {
		if (err) {
			console.log(err);
		} else {
			if (responseData.messages[0]["status"] === "0") {
				console.dir(responseData);
			} else {
				console.log(
					`Message failed with error: ${responseData.messages[0]["error-text"]}`
				);
			}
		}
	}
);
```

> Let's run this, and see if you get a SMS to your mobile phone.

Reviewer: ```suggestion
Let's run this, and see if you get an SMS to your mobile phone.
```

`$ node index.js`

![SMS sent via Vonage on Android](/content/blog/how-to-send-sms-messages-with-node-js-and-express/screenshot-sms.png)

I hope it worked! You have learned how to send an SMS message with the Vonage Node.js SDK.

You can stop right here, or proceed to play with Express.js to be able to take the queries dynamically from POST requests!

### Building a Bare Minimal SMS App with Express.js

<!-- THIS IS HOW FAR YOU'VE GOTTEN.

THE SIMPLE SEND SMS APP WORKS, NOW IT'S TIME TO EXTEND IT AND UPDATE THE TEXT BELOW.

DON'T FORGET TO CREATE A GITHUB REPO WITH RUNNING INSTRUCTIONS. -->

> 

Reviewer: Is this a note for you?

Reviewer 2: Shoot. Yes 😅

> Let’s write a very simple app using [Express](http://expressjs.com/) to send an SMS.

Reviewer: ```suggestion
Let’s write a very simple app using [Express](https://expressjs.com/) to send an SMS.
```

Install Express as a dependency:

```bash
$ npm install express --save
```

In **index.js**, add the following code to start a server that listens on port 3000 for connections:

```javascript
import express from "express";

const { json, urlencoded } = express;

const app = express();

app.use(json());
app.use(
	urlencoded({
		extended: true,
	})
);

app.listen(3000, () => {
	console.log("Server listening at http://localhost:3000");
});
```

> Now, wrap `vonage.message.sendSms()` with the Express `post` route method. Let’s set the `type` to `'unicode'` so you can send some emoji too! Also, print out the success response at the callback.

Reviewer: ```suggestion
Now, wrap `vonage.message.sendSms()` with the Express `post` route method. Let’s set the `type` to `'unicode'` so you can send some emojis too! Also, print out the success response at the callback.
```

```javascript
app.post('/send', (req, res) => {
    // Send SMS
    vonage.message.sendSms(YOUR_VIRTUAL_NUMBER, req.body.toNumber, req.body.message, {type: 'unicode'}, (err, responseData) => {
      if (err) {
          console.log(err);
      } else {
          if(responseData.messages[0]['status'] === "0") {
              console.dir(responseData);
          } else {
              console.log(`Message failed with error: ${responseData.messages[0]['error-text']}`);
          }
      }
  })
});
```

Now, you can try sending an SMS to any mobile phone number (including Google Voice numbers) using your app.

> In this tutorial, we are not going to create an HTML with the form UI where a user can fill out a phone number and message (I will write a full tutorial including a front-end code sometimes!), so let’s pretend we are sending data from a web interface by using [Postman](https://www.getpostman.com/) to make requests to your app. Postman is a good tool to have when you develop apps with REST APIs!

Reviewer: ```suggestion
In this tutorial, we are not going to create the HTML with the form UI where a user can fill out a phone number and message (I will write a full tutorial including the front-end code in the future!), so let’s pretend we are sending data from a web interface by using [Postman](https://www.getpostman.com/) to make requests to your app. Postman is a good tool to have when you develop apps with REST APIs!
```

1. Launch Postman and Select **POST**, and enter _http://localhost:3000/send_.
2. At **Headers**, Set _Content-Type: application/json_

![Send a post request to your app using postman](/content/blog/how-to-send-sms-messages-with-node-js-and-express/postman-headers.png)

3. At **Body**, type a valid JSON with "toNumber" and its value (use your mobile phone number! To receive an SMS message!), also “message” and its value.

![Send a post request to your app using postman](/content/blog/how-to-send-sms-messages-with-node-js-and-express/postman-body.png)

4. Press the blue **Send** button

Once you made a successful POST request to your app, you will get a text message to your phone from the virtual number!

![SMS on Android](/content/blog/how-to-send-sms-messages-with-node-js-and-express/screencast-sms.gif)

> Also, you will see the response printed on your terminal.

Reviewer: ```suggestion
Also, you will see the response printed in your terminal.
```

```shell
{
  'message-count': '1',
  messages: [
    {
      to: '15105551234',
      'message-id': '13000001EBFD617E',
      status: '0',
      'remaining-balance': '5.03151152',
      'message-price': '0.03330000',
      network: 'US_VIRTUAL_BANDWIDTH'
    }
  ]
}
```

You can view [the code sample](https://github.com/nexmo-community/send-sms-nodejs-express/blob/main/index.js) used in this tutorial on GitHub.

> In the next tutorial, you will learn how to receive SMS messages to your virtual number. Stay tuned!

Reviewer: Did we maybe want to link to https://learn.vonage.com/blog/2016/10/27/receive-sms-messages-node-js-express-dr/ or https://learn.vonage.com/blog/2019/09/16/how-to-send-and-receive-sms-messages-with-node-js-and-express-dr/? It makes it seem like aren't any posts yet.

Reviewer 2: Good call, was going to do that once the other post is updated as well. I'll add the link now, as that won't change anyway.
